client/state/comments/reducer.js

```diff
@@ -0,0 +1,266 @@
+
+				if ( commentsTree.getIn( parentPropPath ) !== proposedParent ) {
+					commentsTree.setIn( parentPropPath, proposedParent );
+					commentNodeChanged = true;
+				}
+
+				if ( commentsTree.getIn( dataPropPath ) === undefined ) {
+					commentsTree.setIn( dataPropPath, Immutable.fromJS( comment ) );
+					commentNodeChanged = true;
+				}
+			}
+
+			if ( comment.parent ) {
+				const parentChildrenPath = [ comment.parent.ID, 'children' ];
+				const parentHasChild = commentsTree.getIn( parentChildrenPath ).contains( comment.ID );
+
+				// if parent's children list don't already has that comment, insert it
+				if ( ! parentHasChild ) {
+					commentsTree.updateIn( parentChildrenPath, ( children ) => children.push( comment.ID ) );
+				}
+			}
+
+			// We check here for commentNodeChanged in order to not add the comment if we already saw it
+			if ( commentNodeChanged ) {
+				commentsTree.update( 'fetchedCommentsCount', ( fetchedCommentsCount ) => fetchedCommentsCount + 1 );
+
+				if ( commentsTree.getIn( [ comment.ID, 'parentId' ] ) === null ) {
+					commentsTree.update( 'children', ( children ) => children.push( comment.ID ) );
+				}
+			}
+		} );// forEach comments
+	} ); // withMutation
+
+	return newTree;
+}
+
+/***
+ * Removes comment from the comment tree
+ * @param {Immutable.Map} oldTree Immutable map comments tree instance of the shape Map<id, CommentNode>{ children: List<id> }
+ * @param {Number} commentId the requested comment id for removal
+ * @returns {Immutable.Map} Immutable map comments tree instance of the shape Map<id, CommentNode>{ children: List<id> }
+ */
+function removeComment( oldTree, commentId ) {
+	const newTree = oldTree.withMutations( ( commentsTree ) => {
+		const commentNode = commentsTree.get( commentId );
+
+		if ( commentNode ) {
+			if ( commentNode.get( 'parentId' ) ) {
+				commentsTree.updateIn( [ commentNode.get( 'parentId' ), 'children' ], ( children ) => children.filter( ( childId ) => childId !== commentId ) );
+			}
+
+			commentsTree.delete( commentId );
+			commentsTree.update( 'fetchedCommentsCount', ( fetchedCommentsCount ) => fetchedCommentsCount - 1 );
+		}
+	} );
+
+	return newTree;
+}
+
+/***
+ * Stores comments tree per commentTargetId (siteId, postId)
+ * @param {Immutable.Map} state redux state comments tree
+ * @param {Object} action redux action
+ * @returns {Object} redux state
+ */
+export function tree( state, action ) {
+	switch ( action.type ) {
+		case COMMENTS_RECEIVE:
+			return buildCommentsTree( state, action.comments );
+
+		case COMMENTS_REMOVE_COMMENT:
+			return removeComment( state, action.commentId );
+
+		case DESERIALIZE:
+			// the combined reducer already handled fromJS() for us, the only thing left
+			// is fix the numeric keys, which JS has as strings.
+			return state.mapKeys( ( k ) => ( ! isNaN( k ) ) ? parseInt( k, 10 ) : k );
+			break;
+
+		case SERIALIZE:
+			// no need to do anything here, since the combined reducer will handle it for us.
```

Comment:
Maybe `return state` rather than the break? It makes it a bit clearer if new cases get added below.